{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bitome.core import Bitome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_bitome = Bitome.init_from_file('bitome.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "B=test_bitome.matrix_csc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils import CodonUsage\n",
    "codon_count_gene = CodonUsage.CodonsDict.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### codon count for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.6/dist-packages/Bio/Seq.py:165: BiopythonWarning: Biopython Seq objects now use string comparison. Older versions of Biopython used object comparison. During this transition, please use hash(id(my_seq)) or my_dict[id(my_seq)] if you want the old behaviour, or use hash(str(my_seq)) or my_dict[str(my_seq)] for the new string hashing behaviour.\n",
      "  \"the new string hashing behaviour.\", BiopythonWarning)\n"
     ]
    }
   ],
   "source": [
    "for i in range(0,len(test_bitome.genes)):\n",
    "    \n",
    "    if test_bitome.genes[i].protein and test_bitome.genes[i].transcription_units:\n",
    "#         print (test_bitome.genes[i].gene_type)\n",
    "        for j in range(0, len(test_bitome.genes[i].sequence), 3): \n",
    "            codon = test_bitome.genes[i].sequence[j:j + 3] \n",
    "\n",
    "            if codon in codon_count_gene: \n",
    "                codon_count_gene[codon] += 1 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Codon Usage (RSCU) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# with open('genes_fasta.fasta', 'w') as f:\n",
    "#     for i in range(0, len(test_bitome.genes)):\n",
    "#         if test_bitome.genes[i].protein and test_bitome.genes[i].transcription_units:\n",
    "#             f.write(\"%s\\n\" % (\">\"+test_bitome.genes[i].name + \"\\n\" + test_bitome.genes[i].sequence))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CODON USAGE FROM Bio.SeqUtils but it has been proved to be wrong in some papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from Bio.SeqUtils import CodonUsage as CU\n",
    "# myIndex = CU.CodonAdaptationIndex()\n",
    "# myIndex.generate_index(\"genes_fasta.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# synonymous_codon= CodonUsage.SynonymousCodons.copy()\n",
    "# def codon_usage(c):\n",
    "#     for aa in synonymous_codon:\n",
    "#         codons = synonymous_codon[aa]\n",
    "#         for codon in codons:\n",
    "#             if c == codon:\n",
    "                \n",
    "#                 return (aa, codon,str(myIndex.index[codon]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### codom usage implementation of Bio.SeqUtils by Benjamin Lee (correction!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'test_bitome' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-732f9ea02a64>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mseq\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtest_bitome\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mtest_bitome\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotein\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mtest_bitome\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtranscription_units\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m         \u001b[0mseq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtest_bitome\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgenes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mCAI\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mRSCU\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'test_bitome' is not defined"
     ]
    }
   ],
   "source": [
    "seq=[]\n",
    "for i in range(0, len(test_bitome.genes)):\n",
    "    if test_bitome.genes[i].protein and test_bitome.genes[i].transcription_units:\n",
    "        seq.append(str(test_bitome.genes[i].sequence))\n",
    "from CAI import RSCU\n",
    "rscu = RSCU(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "synonymous_codon= CodonUsage.SynonymousCodons.copy()\n",
    "def codon_usage(c):\n",
    "    for aa in synonymous_codon:\n",
    "        codons = synonymous_codon[aa]\n",
    "        for codon in codons:\n",
    "            if c == codon:\n",
    "                \n",
    "                return (aa, codon,rscu[codon])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GCA'"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairs = {'A':'T','C':'G','G':'C','U':'A'}\n",
    "def codon(anticodon):\n",
    "    \"\"\"returns the codon reverse complement for a given anticodon\"\"\"\n",
    "    complement = ''.join(pairs[c] for c in anticodon)\n",
    "    return complement[::-1]\n",
    "\n",
    "codon('UGC')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCA\n"
     ]
    }
   ],
   "source": [
    "anticodon = 'UGC'\n",
    "complement = ''.join(pairs[c] for c in anticodon)\n",
    "print (complement[::-1])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(('ALA', 'GCC', 1.0778627189256966), ('ALA', 'GCA', 0.8466933624560126), 'GCA')"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codon_usage(\"GCC\"), codon_usage(\"GCA\"),codon('UGC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from __future__ import print_function \n",
    "# import math \n",
    "# from Bio.SeqUtils.CodonUsageIndices import SharpEcoliIndex \n",
    "\n",
    "# synonymous_codon= CodonUsage.SynonymousCodons.copy()\n",
    "# index = SharpEcoliIndex\n",
    " \n",
    "# rcsu = {}\n",
    "# #if test_bitome.genes[i].protein and test_bitome.genes[i].transcription_units:\n",
    "# for aa in synonymous_codon:\n",
    "#     t =[]\n",
    "#     total = 0.0\n",
    "   \n",
    "\n",
    "#     codons = synonymous_codon[aa] \n",
    "  \n",
    "#     for codon in codons: \n",
    "#         total = codon_count_gene[codon]\n",
    "#         #print (total)\n",
    "#         #print ( codon,test_bitome.genes[i].name)\n",
    "#         if not total ==0:\n",
    "#             t.append((codon_count_gene[codon]))\n",
    "#     totalsum= sum(t)\n",
    "  \n",
    "#     for codon in codons:\n",
    "#         total = codon_count_gene[codon]\n",
    "#         if not total ==0:\n",
    "#             denominator = float(total) / len(codons) \n",
    "          \n",
    "#             rcsu[codon]=codon_count_gene[codon] / ((len(codons)** -1) * sum(t))\n",
    "   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Anticodon freq and codon usage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('ACG', '4', 'CGT', ('ARG', 'CGT', 2.3217505396368563)),\n",
       " ('CAA', '1', 'TTG', ('LEU', 'TTG', 0.7685426330863628)),\n",
       " ('CAG', '4', 'CTG', ('LEU', 'CTG', 3.0044368132859116)),\n",
       " ('CAU', '7', 'ATG', ('MET', 'ATG', 1.0)),\n",
       " ('CCA', '1', 'TGG', ('TRP', 'TGG', 1.0)),\n",
       " ('CCC', '1', 'GGG', ('GLY', 'GGG', 0.5974170166856951)),\n",
       " ('CCG', '1', 'CGG', ('ARG', 'CGG', 0.5648622338849622)),\n",
       " ('CCU', '1', 'AGG', ('ARG', 'AGG', 0.10919710500698354)),\n",
       " ('CGA', '1', 'TCG', ('SER', 'TCG', 0.9332221938398247)),\n",
       " ('CGG', '1', 'CCG', ('PRO', 'CCG', 2.1314498451689223)),\n",
       " ('CGU', '1', 'ACG', ('THR', 'ACG', 1.0745769552307647)),\n",
       " ('CUG', '2', 'CAG', ('GLN', 'CAG', 1.3070819454859355)),\n",
       " ('GAA', '2', 'TTC', ('PHE', 'TTC', 0.8520533799348637)),\n",
       " ('GAC', '2', 'GTC', ('VAL', 'GTC', 0.8640496495453855)),\n",
       " ('GAG', '1', 'CTC', ('LEU', 'CTC', 0.6250495985225052)),\n",
       " ('GAU', '4', 'ATC', ('ILE', 'ATC', 1.2675371454299866)),\n",
       " ('GCA', '1', 'TGC', ('CYS', 'TGC', 1.1159420289855073)),\n",
       " ('GCC', '4', 'GGC', ('GLY', 'GGC', 1.6294791088393334)),\n",
       " ('GCU', '1', 'AGC', ('SER', 'AGC', 1.676476519541281)),\n",
       " ('GGA', '2', 'TCC', ('SER', 'TCC', 0.9011978865570509)),\n",
       " ('GGC', '2', 'GCC', ('ALA', 'GCC', 1.0778627189256966)),\n",
       " ('GGG', '1', 'CCC', ('PRO', 'CCC', 0.4858564420166313)),\n",
       " ('GGU', '2', 'ACC', ('THR', 'ACC', 1.7614805846233677)),\n",
       " ('GUA', '3', 'TAC', ('TYR', 'TAC', 0.8646741954007273)),\n",
       " ('GUC', '3', 'GAC', ('ASP', 'GAC', 0.748309971544287)),\n",
       " ('GUG', '1', 'CAC', ('HIS', 'CAC', 0.859661156457919)),\n",
       " ('GUU', '4', 'AAC', ('ASN', 'AAC', 1.1097425301970756)),\n",
       " ('UAA', '1', 'TTA', ('LEU', 'TTA', 0.772654801497695)),\n",
       " ('UAC', '5', 'GTA', ('VAL', 'GTA', 0.6134583251958423)),\n",
       " ('UAG', '1', 'CTA', ('LEU', 'CTA', 0.21556419672036536)),\n",
       " ('UCC', '1', 'GGA', ('GLY', 'GGA', 0.41792795718071457)),\n",
       " ('UCU', '1', 'AGA', ('ARG', 'AGA', 0.19765522495450122)),\n",
       " ('UGA', '1', 'TCA', ('SER', 'TCA', 0.7226845565399767)),\n",
       " ('UGC', '3', 'GCA', ('ALA', 'GCA', 0.8466933624560126)),\n",
       " ('UGG', '1', 'CCA', ('PRO', 'CCA', 0.7567586374865175)),\n",
       " ('UGU', '1', 'ACA', ('THR', 'ACA', 0.5059240681484044)),\n",
       " ('UUC', '4', 'GAA', ('GLU', 'GAA', 1.3844565421437958)),\n",
       " ('UUG', '2', 'CAA', ('GLN', 'CAA', 0.6929180545140645)),\n",
       " ('UUU', '6', 'AAA', ('LYS', 'AAA', 1.5408571327458682))]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ac={}\n",
    "ac1=[]\n",
    "import pandas as pd\n",
    "for i in range(0, len(test_bitome.genes)):\n",
    "    if test_bitome.genes[i].gene_type == 'tRNA':\n",
    "        #print (test_bitome.genes[i].trna_anticodon, test_bitome.genes[i].trna_name, test_bitome.genes[i].name)\n",
    "        if test_bitome.genes[i].trna_anticodon == \"NA1\":\n",
    "                test_bitome.genes[i].trna_anticodon = \"CGG\"\n",
    "        if test_bitome.genes[i].trna_anticodon is not None:\n",
    "            \n",
    "            ac1.append(test_bitome.genes[i].trna_anticodon)\n",
    "            ac.setdefault(test_bitome.genes[i].trna_name, [])\n",
    "            ac[test_bitome.genes[i].trna_name].append(test_bitome.genes[i].trna_anticodon)\n",
    "            #print (ac)\n",
    "        \n",
    "import itertools\n",
    "# from collections import Counter\n",
    "# for elements in ac:\n",
    "#print (\"Anticodon\", \"freq\", \"Codon\", \"Codon Usage\")  \n",
    "res =[]\n",
    "for (key,group) in itertools.groupby(sorted(ac1)): \n",
    "    #print (codon_usage(codon(key)))\n",
    "    #print (key)\n",
    "    anticodon_freq= key,str(len(list(group))), codon(key), codon_usage(codon(key))\n",
    "                                                                      \n",
    "                                                                    \n",
    "    res.append(anticodon_freq)\n",
    "    #print (\", \".join(res))\n",
    "    #df = pd.DataFrame(res)\n",
    "\n",
    "df = pd.DataFrame(res, columns=[\"Anticodon\", \"freq\", \"Codon\", \"Codon Usage\"])\n",
    "df.sort_values('Codon Usage')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CYS': ['TGT', 'TGC'],\n",
       " 'ASP': ['GAT', 'GAC'],\n",
       " 'SER': ['TCT', 'TCG', 'TCA', 'TCC', 'AGC', 'AGT'],\n",
       " 'GLN': ['CAA', 'CAG'],\n",
       " 'MET': ['ATG'],\n",
       " 'ASN': ['AAC', 'AAT'],\n",
       " 'PRO': ['CCT', 'CCG', 'CCA', 'CCC'],\n",
       " 'LYS': ['AAG', 'AAA'],\n",
       " 'THR': ['ACC', 'ACA', 'ACG', 'ACT'],\n",
       " 'PHE': ['TTT', 'TTC'],\n",
       " 'ALA': ['GCA', 'GCC', 'GCG', 'GCT'],\n",
       " 'GLY': ['GGT', 'GGG', 'GGA', 'GGC'],\n",
       " 'ILE': ['ATC', 'ATA', 'ATT'],\n",
       " 'LEU': ['TTA', 'TTG', 'CTC', 'CTT', 'CTG', 'CTA'],\n",
       " 'HIS': ['CAT', 'CAC'],\n",
       " 'ARG': ['CGA', 'CGC', 'CGG', 'CGT', 'AGG', 'AGA'],\n",
       " 'TRP': ['TGG'],\n",
       " 'VAL': ['GTA', 'GTC', 'GTG', 'GTT'],\n",
       " 'GLU': ['GAG', 'GAA'],\n",
       " 'TYR': ['TAT', 'TAC']}"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# synonymous_codon1 = synonymous_codon\n",
    "# del synonymous_codon1[\"STOP\"]\n",
    "synonymous_codon1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "synonymous_codon= CodonUsage.SynonymousCodons.copy()\n",
    "synonymous_codon1 = synonymous_codon\n",
    "del synonymous_codon1[\"STOP\"]\n",
    "result =[]\n",
    "for aa in synonymous_codon1:\n",
    "    codons = synonymous_codon1[aa]\n",
    "    \n",
    "    for c in codons:\n",
    "        #result = codon_usage(codon)\n",
    "        #print (result[1])\n",
    "        anticodon_freq= \"NA\",\"0\", c, codon_usage(c)\n",
    "        for (key,group) in itertools.groupby(sorted(ac1)):\n",
    "            if codon(key) == c:\n",
    "                anticodon_freq= key,str(len(list(group))), codon(key), codon_usage(codon(key))\n",
    "            \n",
    "        \n",
    "        result.append(anticodon_freq)\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Anticodon</th>\n",
       "      <th>freq</th>\n",
       "      <th>Codon</th>\n",
       "      <th>Codon Usage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>UGC</td>\n",
       "      <td>3</td>\n",
       "      <td>GCA</td>\n",
       "      <td>(ALA, GCA, 0.8466933624560126)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>GGC</td>\n",
       "      <td>2</td>\n",
       "      <td>GCC</td>\n",
       "      <td>(ALA, GCC, 1.0778627189256966)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>GCG</td>\n",
       "      <td>(ALA, GCG, 1.4386279982202808)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>GCT</td>\n",
       "      <td>(ALA, GCT, 0.6368159203980099)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>UCU</td>\n",
       "      <td>1</td>\n",
       "      <td>AGA</td>\n",
       "      <td>(ARG, AGA, 0.19765522495450122)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>CCU</td>\n",
       "      <td>1</td>\n",
       "      <td>AGG</td>\n",
       "      <td>(ARG, AGG, 0.10919710500698354)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>CGA</td>\n",
       "      <td>(ARG, CGA, 0.3703390189190333)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>CGC</td>\n",
       "      <td>(ARG, CGC, 2.4361958775976635)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>CCG</td>\n",
       "      <td>1</td>\n",
       "      <td>CGG</td>\n",
       "      <td>(ARG, CGG, 0.5648622338849622)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>ACG</td>\n",
       "      <td>4</td>\n",
       "      <td>CGT</td>\n",
       "      <td>(ARG, CGT, 2.3217505396368563)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GUU</td>\n",
       "      <td>4</td>\n",
       "      <td>AAC</td>\n",
       "      <td>(ASN, AAC, 1.1097425301970756)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>AAT</td>\n",
       "      <td>(ASN, AAT, 0.8902574698029243)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GUC</td>\n",
       "      <td>3</td>\n",
       "      <td>GAC</td>\n",
       "      <td>(ASP, GAC, 0.748309971544287)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>GAT</td>\n",
       "      <td>(ASP, GAT, 1.2516900284557129)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA</td>\n",
       "      <td>1</td>\n",
       "      <td>TGC</td>\n",
       "      <td>(CYS, TGC, 1.1159420289855073)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>TGT</td>\n",
       "      <td>(CYS, TGT, 0.8840579710144928)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>UUG</td>\n",
       "      <td>2</td>\n",
       "      <td>CAA</td>\n",
       "      <td>(GLN, CAA, 0.6929180545140645)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CUG</td>\n",
       "      <td>2</td>\n",
       "      <td>CAG</td>\n",
       "      <td>(GLN, CAG, 1.3070819454859355)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>UUC</td>\n",
       "      <td>4</td>\n",
       "      <td>GAA</td>\n",
       "      <td>(GLU, GAA, 1.3844565421437958)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>GAG</td>\n",
       "      <td>(GLU, GAG, 0.6155434578562042)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>UCC</td>\n",
       "      <td>1</td>\n",
       "      <td>GGA</td>\n",
       "      <td>(GLY, GGA, 0.41792795718071457)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>GCC</td>\n",
       "      <td>4</td>\n",
       "      <td>GGC</td>\n",
       "      <td>(GLY, GGC, 1.6294791088393334)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>CCC</td>\n",
       "      <td>1</td>\n",
       "      <td>GGG</td>\n",
       "      <td>(GLY, GGG, 0.5974170166856951)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>GGT</td>\n",
       "      <td>(GLY, GGT, 1.3551759172942568)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>GUG</td>\n",
       "      <td>1</td>\n",
       "      <td>CAC</td>\n",
       "      <td>(HIS, CAC, 0.859661156457919)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>CAT</td>\n",
       "      <td>(HIS, CAT, 1.1403388435420811)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>ATA</td>\n",
       "      <td>(ILE, ATA, 0.19921528269119446)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>GAU</td>\n",
       "      <td>4</td>\n",
       "      <td>ATC</td>\n",
       "      <td>(ILE, ATC, 1.2675371454299866)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>ATT</td>\n",
       "      <td>(ILE, ATT, 1.5332475718788192)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>UAG</td>\n",
       "      <td>1</td>\n",
       "      <td>CTA</td>\n",
       "      <td>(LEU, CTA, 0.21556419672036536)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>CAG</td>\n",
       "      <td>4</td>\n",
       "      <td>CTG</td>\n",
       "      <td>(LEU, CTG, 3.0044368132859116)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>CTT</td>\n",
       "      <td>(LEU, CTT, 0.6137519568871607)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>UAA</td>\n",
       "      <td>1</td>\n",
       "      <td>TTA</td>\n",
       "      <td>(LEU, TTA, 0.772654801497695)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>CAA</td>\n",
       "      <td>1</td>\n",
       "      <td>TTG</td>\n",
       "      <td>(LEU, TTG, 0.7685426330863628)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>UUU</td>\n",
       "      <td>6</td>\n",
       "      <td>AAA</td>\n",
       "      <td>(LYS, AAA, 1.5408571327458682)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>AAG</td>\n",
       "      <td>(LYS, AAG, 0.4591428672541317)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CAU</td>\n",
       "      <td>7</td>\n",
       "      <td>ATG</td>\n",
       "      <td>(MET, ATG, 1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>GAA</td>\n",
       "      <td>2</td>\n",
       "      <td>TTC</td>\n",
       "      <td>(PHE, TTC, 0.8520533799348637)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>TTT</td>\n",
       "      <td>(PHE, TTT, 1.1479466200651363)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>UGG</td>\n",
       "      <td>1</td>\n",
       "      <td>CCA</td>\n",
       "      <td>(PRO, CCA, 0.7567586374865175)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GGG</td>\n",
       "      <td>1</td>\n",
       "      <td>CCC</td>\n",
       "      <td>(PRO, CCC, 0.4858564420166313)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CGG</td>\n",
       "      <td>1</td>\n",
       "      <td>CCG</td>\n",
       "      <td>(PRO, CCG, 2.1314498451689223)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>CCT</td>\n",
       "      <td>(PRO, CCT, 0.6259350753279287)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GCU</td>\n",
       "      <td>1</td>\n",
       "      <td>AGC</td>\n",
       "      <td>(SER, AGC, 1.676476519541281)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>AGT</td>\n",
       "      <td>(SER, AGT, 0.8943412968365578)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>UGA</td>\n",
       "      <td>1</td>\n",
       "      <td>TCA</td>\n",
       "      <td>(SER, TCA, 0.7226845565399767)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GGA</td>\n",
       "      <td>2</td>\n",
       "      <td>TCC</td>\n",
       "      <td>(SER, TCC, 0.9011978865570509)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CGA</td>\n",
       "      <td>1</td>\n",
       "      <td>TCG</td>\n",
       "      <td>(SER, TCG, 0.9332221938398247)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>TCT</td>\n",
       "      <td>(SER, TCT, 0.8720775466853095)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>UGU</td>\n",
       "      <td>1</td>\n",
       "      <td>ACA</td>\n",
       "      <td>(THR, ACA, 0.5059240681484044)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>GGU</td>\n",
       "      <td>2</td>\n",
       "      <td>ACC</td>\n",
       "      <td>(THR, ACC, 1.7614805846233677)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CGU</td>\n",
       "      <td>1</td>\n",
       "      <td>ACG</td>\n",
       "      <td>(THR, ACG, 1.0745769552307647)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>ACT</td>\n",
       "      <td>(THR, ACT, 0.6580183919974631)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>CCA</td>\n",
       "      <td>1</td>\n",
       "      <td>TGG</td>\n",
       "      <td>(TRP, TGG, 1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>GUA</td>\n",
       "      <td>3</td>\n",
       "      <td>TAC</td>\n",
       "      <td>(TYR, TAC, 0.8646741954007273)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>TAT</td>\n",
       "      <td>(TYR, TAT, 1.1353258045992727)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>UAC</td>\n",
       "      <td>5</td>\n",
       "      <td>GTA</td>\n",
       "      <td>(VAL, GTA, 0.6134583251958423)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>GAC</td>\n",
       "      <td>2</td>\n",
       "      <td>GTC</td>\n",
       "      <td>(VAL, GTC, 0.8640496495453855)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>GTG</td>\n",
       "      <td>(VAL, GTG, 1.4926545580799861)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>GTT</td>\n",
       "      <td>(VAL, GTT, 1.029837467178786)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Anticodon freq Codon                      Codon Usage\n",
       "27       UGC    3   GCA   (ALA, GCA, 0.8466933624560126)\n",
       "28       GGC    2   GCC   (ALA, GCC, 1.0778627189256966)\n",
       "29        NA    0   GCG   (ALA, GCG, 1.4386279982202808)\n",
       "30        NA    0   GCT   (ALA, GCT, 0.6368159203980099)\n",
       "51       UCU    1   AGA  (ARG, AGA, 0.19765522495450122)\n",
       "50       CCU    1   AGG  (ARG, AGG, 0.10919710500698354)\n",
       "46        NA    0   CGA   (ARG, CGA, 0.3703390189190333)\n",
       "47        NA    0   CGC   (ARG, CGC, 2.4361958775976635)\n",
       "48       CCG    1   CGG   (ARG, CGG, 0.5648622338849622)\n",
       "49       ACG    4   CGT   (ARG, CGT, 2.3217505396368563)\n",
       "13       GUU    4   AAC   (ASN, AAC, 1.1097425301970756)\n",
       "14        NA    0   AAT   (ASN, AAT, 0.8902574698029243)\n",
       "3        GUC    3   GAC    (ASP, GAC, 0.748309971544287)\n",
       "2         NA    0   GAT   (ASP, GAT, 1.2516900284557129)\n",
       "1        GCA    1   TGC   (CYS, TGC, 1.1159420289855073)\n",
       "0         NA    0   TGT   (CYS, TGT, 0.8840579710144928)\n",
       "10       UUG    2   CAA   (GLN, CAA, 0.6929180545140645)\n",
       "11       CUG    2   CAG   (GLN, CAG, 1.3070819454859355)\n",
       "58       UUC    4   GAA   (GLU, GAA, 1.3844565421437958)\n",
       "57        NA    0   GAG   (GLU, GAG, 0.6155434578562042)\n",
       "33       UCC    1   GGA  (GLY, GGA, 0.41792795718071457)\n",
       "34       GCC    4   GGC   (GLY, GGC, 1.6294791088393334)\n",
       "32       CCC    1   GGG   (GLY, GGG, 0.5974170166856951)\n",
       "31        NA    0   GGT   (GLY, GGT, 1.3551759172942568)\n",
       "45       GUG    1   CAC    (HIS, CAC, 0.859661156457919)\n",
       "44        NA    0   CAT   (HIS, CAT, 1.1403388435420811)\n",
       "36        NA    0   ATA  (ILE, ATA, 0.19921528269119446)\n",
       "35       GAU    4   ATC   (ILE, ATC, 1.2675371454299866)\n",
       "37        NA    0   ATT   (ILE, ATT, 1.5332475718788192)\n",
       "43       UAG    1   CTA  (LEU, CTA, 0.21556419672036536)\n",
       "..       ...  ...   ...                              ...\n",
       "42       CAG    4   CTG   (LEU, CTG, 3.0044368132859116)\n",
       "41        NA    0   CTT   (LEU, CTT, 0.6137519568871607)\n",
       "38       UAA    1   TTA    (LEU, TTA, 0.772654801497695)\n",
       "39       CAA    1   TTG   (LEU, TTG, 0.7685426330863628)\n",
       "20       UUU    6   AAA   (LYS, AAA, 1.5408571327458682)\n",
       "19        NA    0   AAG   (LYS, AAG, 0.4591428672541317)\n",
       "12       CAU    7   ATG                  (MET, ATG, 1.0)\n",
       "26       GAA    2   TTC   (PHE, TTC, 0.8520533799348637)\n",
       "25        NA    0   TTT   (PHE, TTT, 1.1479466200651363)\n",
       "17       UGG    1   CCA   (PRO, CCA, 0.7567586374865175)\n",
       "18       GGG    1   CCC   (PRO, CCC, 0.4858564420166313)\n",
       "16       CGG    1   CCG   (PRO, CCG, 2.1314498451689223)\n",
       "15        NA    0   CCT   (PRO, CCT, 0.6259350753279287)\n",
       "8        GCU    1   AGC    (SER, AGC, 1.676476519541281)\n",
       "9         NA    0   AGT   (SER, AGT, 0.8943412968365578)\n",
       "6        UGA    1   TCA   (SER, TCA, 0.7226845565399767)\n",
       "7        GGA    2   TCC   (SER, TCC, 0.9011978865570509)\n",
       "5        CGA    1   TCG   (SER, TCG, 0.9332221938398247)\n",
       "4         NA    0   TCT   (SER, TCT, 0.8720775466853095)\n",
       "22       UGU    1   ACA   (THR, ACA, 0.5059240681484044)\n",
       "21       GGU    2   ACC   (THR, ACC, 1.7614805846233677)\n",
       "23       CGU    1   ACG   (THR, ACG, 1.0745769552307647)\n",
       "24        NA    0   ACT   (THR, ACT, 0.6580183919974631)\n",
       "52       CCA    1   TGG                  (TRP, TGG, 1.0)\n",
       "60       GUA    3   TAC   (TYR, TAC, 0.8646741954007273)\n",
       "59        NA    0   TAT   (TYR, TAT, 1.1353258045992727)\n",
       "53       UAC    5   GTA   (VAL, GTA, 0.6134583251958423)\n",
       "54       GAC    2   GTC   (VAL, GTC, 0.8640496495453855)\n",
       "55        NA    0   GTG   (VAL, GTG, 1.4926545580799861)\n",
       "56        NA    0   GTT    (VAL, GTT, 1.029837467178786)\n",
       "\n",
       "[61 rows x 4 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(result, columns=[\"Anticodon\", \"freq\", \"Codon\", \"Codon Usage\"]).sort_values('Codon Usage')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
